File: apps/server/src/routes/channels.ts
/** Channels, membership, and DM creation. */

import type { FastifyInstance } from 'fastify';
import { z } from 'zod';
import {
  createChannelSchema,
  createDmSchema,
  membershipUpdateSchema,
  updateChannelSchema,
} from '@blob/shared';
import { query, queryOne } from '../db/pool.ts';
import { requireUser } from '../lib/auth.ts';
import { forbidden, notFound } from '../lib/errors.ts';
import * as channels from '../services/channels.ts';
import * as messages from '../services/messages.ts';
import * as hub from '../realtime/hub.ts';

const idParam = z.object({ id: z.string().uuid() });

export async function channelRoutes(app: FastifyInstance): Promise<void> {
  app.get('/api/channels', async (req) => {
    const user = requireUser(req);
    return { channels: await channels.listForUser(user.id, user.workspaceId) };
  });

  app.post('/api/channels', async (req) => {
    const user = requireUser(req);
    const input = createChannelSchema.parse(req.body);

    const channelId = await channels.createChannel({
      workspaceId: user.workspaceId,
      createdBy: user.id,
      name: input.name,
      kind: input.kind,
      topic: input.topic ?? null,
      description: input.description ?? null,
      memberIds: input.memberIds,
    });

    const channel = await channels.getForUser(channelId, user.id);
    if (!channel) throw notFound('Could not create that channel.');

    // Public channels appear in everyone's browser; private ones only for members.
    if (input.kind === 'public') {
      hub.toAll({ t: 'channel.created', channel });
    } else {
      hub.toUsers(await channels.memberIds(channelId), { t: 'channel.created', channel });
    }
    return { channel };
  });

  app.get('/api/channels/:id', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    await channels.assertChannelAccess(user.id, id);
    const channel = await channels.getForUser(id, user.id);
    if (!channel) throw notFound('That channel no longer exists.');
    return { channel };
  });

  app.patch('/api/channels/:id', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    const input = updateChannelSchema.parse(req.body);
    const access = await channels.assertChannelAccess(user.id, id, {
      requireMember: true,
      requireWritable: true,
    });
    if (access.kind === 'dm' || access.kind === 'group_dm') {
      throw forbidden('Direct messages have no channel settings.');
    }

    await query(
      `UPDATE channels
          SET name = COALESCE($2, name),
              topic = CASE WHEN $3::boolean THEN $4 ELSE topic END,
              description = CASE WHEN $5::boolean THEN $6 ELSE description END
        WHERE id = $1`,
      [
        id,
        input.name ?? null,
        input.topic !== undefined,
        input.topic ?? null,
        input.description !== undefined,
        input.description ?? null,
      ],
    );

    const channel = await channels.getForUser(id, user.id);
    if (channel) hub.toChannel(id, { t: 'channel.updated', channel });
    return { channel };
  });

  app.post('/api/channels/:id/archive', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    const access = await channels.assertChannelAccess(user.id, id, { requireMember: true });
    if (access.kind === 'dm' || access.kind === 'group_dm') {
      throw forbidden('Direct messages cannot be archived.');
    }
    await query('UPDATE channels SET archived_at = now() WHERE id = $1', [id]);
    hub.toChannel(id, { t: 'channel.archived', channelId: id });
    return { ok: true };
  });

  app.post('/api/channels/:id/join', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    const access = await channels.assertChannelAccess(user.id, id, { requireWritable: true });
    if (access.kind !== 'public') throw forbidden('That channel is invitation-only.');

    await channels.join(id, user.id);
    hub.toChannel(id, { t: 'member.joined', channelId: id, userId: user.id });
    // Existing sockets for this user need to start receiving the channel's events.
    for (const conn of hub.connectionsForUser(user.id)) hub.subscribeChannels(conn, [id]);

    const channel = await channels.getForUser(id, user.id);
    return { channel };
  });

  app.post('/api/channels/:id/leave', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    const access = await channels.assertChannelAccess(user.id, id, { requireMember: true });
    if (access.kind === 'dm' || access.kind === 'group_dm') {
      throw forbidden('You cannot leave a direct message.');
    }
    await channels.leave(id, user.id);
    for (const conn of hub.connectionsForUser(user.id)) hub.unsubscribeChannel(conn, id);
    hub.toChannel(id, { t: 'member.left', channelId: id, userId: user.id });
    return { ok: true };
  });

  app.post('/api/channels/:id/members', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    const { userIds } = z.object({ userIds: z.array(z.string().uuid()).min(1).max(50) }).parse(
      req.body,
    );
    const access = await channels.assertChannelAccess(user.id, id, {
      requireMember: true,
      requireWritable: true,
    });
    if (access.kind === 'dm' || access.kind === 'group_dm') {
      throw forbidden('Start a new group message instead of adding people to this one.');
    }

    for (const userId of userIds) {
      await channels.join(id, userId);
      hub.toChannel(id, { t: 'member.joined', channelId: id, userId });
      for (const conn of hub.connectionsForUser(userId)) hub.subscribeChannels(conn, [id]);
      const channel = await channels.getForUser(id, userId);
      if (channel) hub.toUsers([userId], { t: 'channel.created', channel });
    }
    return { ok: true };
  });

  app.get('/api/channels/:id/members', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    await channels.assertChannelAccess(user.id, id);
    return { userIds: await channels.memberIds(id) };
  });

  /** Per-user channel settings: notification level and starring. */
  app.patch('/api/channels/:id/membership', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    const input = membershipUpdateSchema.parse(req.body);
    await channels.assertChannelAccess(user.id, id, { requireMember: true });

    await query(
      `UPDATE channel_members
          SET notify_level = COALESCE($3, notify_level),
              is_starred   = COALESCE($4, is_starred)
        WHERE channel_id = $1 AND user_id = $2`,
      [id, user.id, input.notifyLevel ?? null, input.isStarred ?? null],
    );

    const channel = await channels.getForUser(id, user.id);
    if (channel) hub.toUsers([user.id], { t: 'channel.updated', channel });
    return { channel };
  });

  app.get('/api/channels/:id/pins', async (req) => {
    const user = requireUser(req);
    const { id } = idParam.parse(req.params);
    await channels.assertChannelAccess(user.id, id);
    return { messages: await messages.listPinned(id) };
  });

  /** Open (or reopen) a DM. Idempotent: the same member set returns the same channel. */
  app.post('/api/dms', async (req) => {
    const user = requireUser(req);
    const { userIds } = createDmSchema.parse(req.body);

    const members = [...new Set([user.id, ...userIds])];
    const valid = await queryOne<{ count: number }>(
      `SELECT count(*)::int AS count FROM users
        WHERE id = ANY($1::uuid[]) AND workspace_id = $2 AND deactivated_at IS NULL`,
      [members, user.workspaceId],
    );
    if ((valid?.count ?? 0) !== members.length) throw notFound('One of those people is unavailable.');

    const { channelId, created } = await channels.findOrCreateDm(user.workspaceId, members);
    const channel = await channels.getForUser(channelId, user.id);

    if (created) {
      for (const memberId of members) {
        for (const conn of hub.connectionsForUser(memberId)) {
          hub.subscribeChannels(conn, [channelId]);
        }
        const view = await channels.getForUser(channelId, memberId);
        if (view) hub.toUsers([memberId], { t: 'channel.created', channel: view });
      }
    }
    return { channel };
  });
}

File: apps/api/src/blob_api/routers/channels.py
"""Channels, membership, and DM creation."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from ..db.engine import session_scope, transaction
from ..lib.auth import SessionUser, current_user
from ..lib.errors import channel_gone, forbidden, not_found
from ..lib.ids import IdParam
from ..lib.queue import enqueue, fire_and_forget
from ..plugins import events as plugin_events
from ..realtime import hub
from ..schemas.base import CamelModel, OkOut
from ..schemas.models import BrowsableChannel, ChannelWithState, MembersOut, MessagesOut
from ..schemas.requests import (
    AddMembersInput,
    CreateChannelInput,
    CreateDmInput,
    MembershipUpdateInput,
    UpdateChannelInput,
)
from ..services import channels as channel_service
from ..services import messages as message_service
from ..services import users as user_service
from ..services.serialize import channel_event, membership_event

router = APIRouter(tags=["channels"])


class ChannelsOut(CamelModel):
    channels: list[ChannelWithState]


class ChannelOut(CamelModel):
    channel: ChannelWithState | None = None


@router.get("/api/channels", response_model=ChannelsOut)
async def list_channels(user: SessionUser = Depends(current_user)) -> ChannelsOut:
    async with session_scope() as session:
        channels = await channel_service.list_for_user(session, user.id, user.workspace_id)
    return ChannelsOut(channels=channels)


class BrowseOut(CamelModel):
    channels: list[BrowsableChannel]


@router.get("/api/channels/browse", response_model=BrowseOut)
async def browse_channels(
    q: str = Query("", max_length=100),
    archived: bool = False,
    user: SessionUser = Depends(current_user),
) -> BrowseOut:
    """The channel directory.

    Public channels only, so nothing here can reveal that a private channel exists —
    the same reason opening one you are not in answers 404 rather than 403.
    """
    async with session_scope() as session:
        channels = await channel_service.browse(
            session, user.id, user.workspace_id, query=q, include_archived=archived
        )
    return BrowseOut(channels=channels)


@router.post("/api/channels", response_model=ChannelOut)
async def create_channel(
    payload: CreateChannelInput, user: SessionUser = Depends(current_user)
) -> ChannelOut:
    async with transaction() as (session, after):
        channel_id = await channel_service.create_channel(
            session,
            workspace_id=user.workspace_id,
            created_by=user.id,
            name=payload.name,
            kind=payload.kind,
            topic=payload.topic,
            description=payload.description,
            extra_member_ids=payload.member_ids,
        )
        channel = await channel_service.get_for_user(session, channel_id, user.id)
        if channel is None:
            raise not_found("Could not create that channel.")
        members = await channel_service.member_ids(session, channel_id)
        views = {
            member_id: await channel_service.get_for_user(session, channel_id, member_id)
            for member_id in members
        }
        # Catalogued in every app's manifest since the beginning; emitted since now.
        # Scoped by channel, so a private channel is announced only to apps invited in.
        await plugin_events.emit(
            session,
            workspace_id=user.workspace_id,
            event="channel.created",
            channel_id=channel_id,
            payload={"channelId": channel_id, "name": payload.name, "kind": payload.kind},
        )

        def broadcast() -> None:
            # Public channels appear in everyone's browser; private ones only for members.
            if payload.kind == "public":
                hub.to_workspace(user.workspace_id, channel_event("channel.created", channel))
            else:
                hub.to_users(members, channel_event("channel.created", channel))
            # Then, to each member alone, their own standing in it. A public channel's
            # arrival reaches the whole workspace and almost nobody there is in it, so
            # the membership half cannot ride the same frame.
            for member_id, view in views.items():
                if view is not None:
                    hub.to_users([member_id], membership_event(view))
            hub.subscribe_users(members, [channel_id])
            fire_and_forget(enqueue("deliver_plugin_events"))

        after.add(broadcast)

    return ChannelOut(channel=channel)


@router.get("/api/channels/{channel_id}", response_model=ChannelOut)
async def get_channel(channel_id: IdParam, user: SessionUser = Depends(current_user)) -> ChannelOut:
    async with session_scope() as session:
        await channel_service.assert_channel_access(session, user.id, channel_id)
        channel = await channel_service.get_for_user(session, channel_id, user.id)
    if channel is None:
        raise channel_gone()
    return ChannelOut(channel=channel)


@router.patch("/api/channels/{channel_id}", response_model=ChannelOut)
async def update_channel(
    channel_id: IdParam,
    payload: UpdateChannelInput,
    user: SessionUser = Depends(current_user),
) -> ChannelOut:
    given = payload.model_fields_set

    async with transaction() as (session, after):
        access = await channel_service.assert_channel_access(
            session, user.id, channel_id, require_member=True, require_writable=True
        )
        if access.kind in ("dm", "group_dm"):
            raise forbidden("Direct messages have no channel settings.")

        await channel_service.update_settings(
            session,
            channel_id,
            name=payload.name,
            topic=payload.topic,
            description=payload.description,
            nudge_unanswered=payload.nudge_unanswered,
            given=given,
        )
        channel = await channel_service.get_for_user(session, channel_id, user.id)
        if channel is not None:
            after.add(lambda: hub.to_channel(channel_id, channel_event("channel.updated", channel)))

    return ChannelOut(channel=channel)


@router.post("/api/channels/{channel_id}/archive", response_model=OkOut)
async def archive_channel(channel_id: IdParam, user: SessionUser = Depends(current_user)) -> OkOut:
    """Close a channel. Admins only — the client has always said so; this enforces it."""
    async with transaction() as (session, after):
        access = await channel_service.assert_channel_access(
            session, user.id, channel_id, require_member=True
        )
        if access.kind in ("dm", "group_dm"):
            raise forbidden("Direct messages cannot be archived.")
        if not user.is_admin:
            raise forbidden("Only an admin can archive a channel.")

        await channel_service.set_archived(session, channel_id, archived=True)
        after.add(
            lambda: hub.to_channel(channel_id, {"t": "channel.archived", "channelId": channel_id})
        )
    return OkOut()


@router.post("/api/channels/{channel_id}/unarchive", response_model=ChannelOut)
async def unarchive_channel(
    channel_id: IdParam, user: SessionUser = Depends(current_user)
) -> ChannelOut:
    """Open an archived channel again.

    Archiving had no undo anywhere in the product — no route, no command, no console
    control, and nothing that set `archived_at` back to null. That made a reversible
    decision permanent by omission: the history was intact the whole time and simply
    unreachable for writing.
    """
    async with transaction() as (session, after):
        access = await channel_service.assert_channel_access(session, user.id, channel_id)
        if access.kind in ("dm", "group_dm"):
            raise forbidden("Direct messages cannot be archived.")
        if not user.is_admin:
            raise forbidden("Only an admin can reopen a channel.")

        await channel_service.set_archived(session, channel_id, archived=False)
        channel = await channel_service.get_for_user(session, channel_id, user.id)
        if channel is not None:
            after.add(lambda: hub.to_channel(channel_id, channel_event("channel.updated", channel)))
    if channel is None:
        raise not_found("That channel is gone.")
    return ChannelOut(channel=channel)


@router.post("/api/channels/{channel_id}/join", response_model=ChannelOut)
async def join_channel(
    channel_id: IdParam, user: SessionUser = Depends(current_user)
) -> ChannelOut:
    async with transaction() as (session, after):
        access = await channel_service.assert_channel_access(
            session, user.id, channel_id, require_writable=True
        )
        if access.kind != "public":
            raise forbidden("That channel is invitation-only.")

        await channel_service.join(session, channel_id, user.id)
        channel = await channel_service.get_for_user(session, channel_id, user.id)
        # People joining were never announced to apps — only bots were (bot_api), so an
        # app subscribed to the catalogued member.joined heard about robots and nobody
        # else. Same event, same scoping, the missing half of the population.
        await plugin_events.emit(
            session,
            workspace_id=user.workspace_id,
            event="member.joined",
            channel_id=channel_id,
            payload={"channelId": channel_id, "userId": user.id},
        )

        def broadcast() -> None:
            hub.to_channel(
                channel_id, {"t": "member.joined", "channelId": channel_id, "userId": user.id}
            )
            # Existing sockets need to start receiving the channel's events —
            # wherever they are held; the join may have landed on a sibling process.
            hub.subscribe_users([user.id], [channel_id])
            fire_and_forget(enqueue("deliver_plugin_events"))

        after.add(broadcast)

    return ChannelOut(channel=channel)


@router.post("/api/channels/{channel_id}/leave", response_model=OkOut)
async def leave_channel(channel_id: IdParam, user: SessionUser = Depends(current_user)) -> OkOut:
    async with transaction() as (session, after):
        access = await channel_service.assert_channel_access(
            session, user.id, channel_id, require_member=True
        )
        if access.kind in ("dm", "group_dm"):
            raise forbidden("You cannot leave a direct message.")

        await channel_service.leave(session, channel_id, user.id)
        await plugin_events.emit(
            session,
            workspace_id=user.workspace_id,
            event="member.left",
            channel_id=channel_id,
            payload={"channelId": channel_id, "userId": user.id},
        )

        def broadcast() -> None:
            hub.unsubscribe_users([user.id], [channel_id])
            hub.to_channel(
                channel_id, {"t": "member.left", "channelId": channel_id, "userId": user.id}
            )
            fire_and_forget(enqueue("deliver_plugin_events"))

        after.add(broadcast)
    return OkOut()


@router.post("/api/channels/{channel_id}/members", response_model=OkOut)
async def add_members(
    channel_id: IdParam, payload: AddMembersInput, user: SessionUser = Depends(current_user)
) -> OkOut:
    async with transaction() as (session, after):
        access = await channel_service.assert_channel_access(
            session, user.id, channel_id, require_member=True, require_writable=True
        )
        if access.kind in ("dm", "group_dm"):
            raise forbidden("Start a new group message instead of adding people to this one.")

        await channel_service.add_members(session, channel_id, payload.user_ids)
        for member_id in payload.user_ids:
            await plugin_events.emit(
                session,
                workspace_id=user.workspace_id,
                event="member.joined",
                channel_id=channel_id,
                payload={"channelId": channel_id, "userId": member_id},
            )
        views = {
            member_id: await channel_service.get_for_user(session, channel_id, member_id)
            for member_id in payload.user_ids
        }

        def broadcast() -> None:
            for member_id, view in views.items():
                hub.to_channel(
                    channel_id,
                    {"t": "member.joined", "channelId": channel_id, "userId": member_id},
                )
                hub.subscribe_users([member_id], [channel_id])
                if view is not None:
                    hub.to_users([member_id], channel_event("channel.created", view))
                    hub.to_users([member_id], membership_event(view))
            fire_and_forget(enqueue("deliver_plugin_events"))

        after.add(broadcast)
    return OkOut()


@router.get("/api/channels/{channel_id}/members", response_model=MembersOut)
async def list_members(
    channel_id: IdParam, user: SessionUser = Depends(current_user)
) -> MembersOut:
    async with session_scope() as session:
        await channel_service.assert_channel_access(session, user.id, channel_id)
        ids = await channel_service.member_ids(session, channel_id)
    return MembersOut(user_ids=ids)


@router.patch("/api/channels/{channel_id}/membership", response_model=ChannelOut)
async def update_membership(
    channel_id: IdParam,
    payload: MembershipUpdateInput,
    user: SessionUser = Depends(current_user),
) -> ChannelOut:
    """Per-user channel settings: notification level and starring."""
    async with transaction() as (session, after):
        await channel_service.assert_channel_access(
            session, user.id, channel_id, require_member=True
        )
        await channel_service.update_membership(
            session,
            channel_id,
            user.id,
            notify_level=payload.notify_level,
            is_starred=payload.is_starred,
        )
        channel = await channel_service.get_for_user(session, channel_id, user.id)
        if channel is not None:
            # How loud a channel is for one person is nobody else's business, and the
            # channel itself did not change.
            after.add(lambda: hub.to_users([user.id], membership_event(channel)))
    return ChannelOut(channel=channel)


@router.get("/api/channels/{channel_id}/pins", response_model=MessagesOut)
async def list_pins(channel_id: IdParam, user: SessionUser = Depends(current_user)) -> MessagesOut:
    async with session_scope() as session:
        await channel_service.assert_channel_access(session, user.id, channel_id)
        messages = await message_service.list_pinned(session, channel_id)
    return MessagesOut(messages=messages)


@router.post("/api/dms", response_model=ChannelOut)
async def open_dm(payload: CreateDmInput, user: SessionUser = Depends(current_user)) -> ChannelOut:
    """Open (or reopen) a DM. Idempotent: the same member set returns the same channel."""
    members = list(dict.fromkeys([user.id, *payload.user_ids]))

    async with transaction() as (session, after):
        if not await user_service.all_active(session, user.workspace_id, members):
            raise not_found("One of those people is unavailable.")

        channel_id, created = await channel_service.find_or_create_dm(
            session, user.workspace_id, members
        )
        channel = await channel_service.get_for_user(session, channel_id, user.id)

        if created:
            views = {
                member_id: await channel_service.get_for_user(session, channel_id, member_id)
                for member_id in members
            }

            def broadcast() -> None:
                for member_id, view in views.items():
                    hub.subscribe_users([member_id], [channel_id])
                    if view is not None:
                        hub.to_users([member_id], channel_event("channel.created", view))
                        hub.to_users([member_id], membership_event(view))

            after.add(broadcast)

    return ChannelOut(channel=channel)


__all__ = ["router"]
